### PR TITLE
[BROWSEUI] Multithreaded auto-completion

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Improve auto-completion without timeout and jerky of typing.
JIRA issue: [CORE-9281](https://jira.reactos.org/browse/CORE-9281)

## Proposed changes

- Make auto-completion multi-threaded.

## TODO

- [x] Do tests.
- [ ] Quick typing.
- [ ] Large number of files search.

### CODE REVIEW

**Comment by katahiromz** [comment]
BEFORE:
before
AFTER:
after

File: dll/win32/browseui/CAutoComplete.cpp [hunk 1]
```diff
@@ -53,7 +53,7 @@ static const PREFIX_INFO s_prefixes[] =
     { L"www.", 4 },
 };
 
-static inline BOOL DropPrefix(const CStringW& str, CStringW& strBody)
+static BOOL DropPrefix(const CStringW& str, CStringW& strBody)
```
Comment (learn-more) [inline dll/win32/browseui/CAutoComplete.cpp:56]: why did you remove inline?

Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp:56]: I thought it is too large to be embedded.

**Comment by katahiromz** [comment]
Check time. I will merge 5 days later.

File: dll/win32/browseui/CAutoComplete.cpp [hunk 2]
```diff
@@ -2116,3 +1979,189 @@ LRESULT CAutoComplete::OnVScroll(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &
...
+    }
+
+    DoSort(pThread->m_outerList);
+    DoUniqueAndTrim(pThread->m_outerList);
+
+    if (m_pThread || !m_hThread ||
+        !::PostMessageW(m_hWnd, AUTOCOMP_FINISH, 0, (LPARAM)pThread))
+    {
+        delete pThread;
+    }
```
Comment (learn-more) [inline dll/win32/browseui/CAutoComplete.cpp:2040]: else pThread is leaked?

File: dll/win32/browseui/CAutoComplete.cpp [hunk 3]
```diff
@@ -2116,3 +1979,189 @@ LRESULT CAutoComplete::OnVScroll(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &
...
+        CloseHandle(m_hThread);
+        bDoStart = TRUE;
+    }
+    else if (!m_hThread)
+    {
+        bDoStart = TRUE;
+    }
+
+    if (bDoStart)
+        m_hThread = (HANDLE)_beginthreadex(NULL, 0, ::AutoCompThreadProc, this, 0, NULL);
```
Comment (learn-more) [inline dll/win32/browseui/CAutoComplete.cpp:2107]: I am not sure that this part always leads to a running thread that will process the autocomplete item.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 2 above)
Comment (learn-more) [reply dll/win32/browseui/CAutoComplete.cpp:2040]: oh, it's deleted while handling the AUTOCOMP_FINISH

File: dll/win32/browseui/CAutoComplete.cpp [hunk 4]
```diff
@@ -2116,3 +1979,189 @@ LRESULT CAutoComplete::OnVScroll(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &
...
+// AUTOCOMP_FINISH
+LRESULT CAutoComplete::OnAutoCompFinish(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    PAC_THREAD pThread = reinterpret_cast<PAC_THREAD>(lParam);
+    if (m_pThread == NULL)
+    {
+        FinishCompletion(pThread);
+        delete pThread;
+    }
+    CloseHandle(m_hThread);
```
Comment (learn-more) [inline dll/win32/browseui/CAutoComplete.cpp:2121]: does pThread not need deletion in this case?

File: dll/win32/browseui/CAutoComplete.cpp [hunk 5]
```diff
@@ -2116,3 +1979,189 @@ LRESULT CAutoComplete::OnVScroll(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &
...
+        return;
+
+    // set inner list
+    m_innerList = pThread->m_innerList;
+
+    if (m_pThread || !m_hThread)
+        return;
+
+    // set the items of the virtual listview
+    m_outerList = pThread->m_outerList; // FIXME: We need more speed!
```
Comment (learn-more) [inline dll/win32/browseui/CAutoComplete.cpp:2144]: Virtual listbox is faaaaaast, (but not very easy to implement, and needs special handling in the listbox owner):
https://docs.microsoft.com/en-us/cpp/mfc/virtual-list-controls?view=msvc-160

Comment (HBelusca) [reply dll/win32/browseui/CAutoComplete.cpp:2144]: Note: It's MFC (and not ATL); we don't use MFC.

File: dll/win32/browseui/CAutoComplete.cpp [hunk 6]
```diff
@@ -2116,3 +1979,189 @@ LRESULT CAutoComplete::OnVScroll(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &
...
+        bDoStart = TRUE;
+    }
+    else if (!m_hThread)
+    {
+        bDoStart = TRUE;
+    }
+
+    if (bDoStart)
+        m_hThread = (HANDLE)_beginthreadex(NULL, 0, ::AutoCompThreadProc, this, 0, NULL);
+
```
Comment (katahiromz) [inline dll/win32/browseui/CAutoComplete.cpp:2108]: Cc @ThFabba

File: dll/win32/browseui/CAutoComplete.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp:2121]: Cc @ThFabba

**Comment by katahiromz** [comment]
The problem is that, if the user quickly typed many characters at the folder that has many (10000+) items, then auto-completion won't work.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 6 above)
Comment (learn-more) [reply dll/win32/browseui/CAutoComplete.cpp:2108]: If you have a specific question, it's better to ask a question instead of just cc'ing someone...

File: dll/win32/browseui/CAutoComplete.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp:2120]: Moved.

File: dll/win32/browseui/CAutoComplete.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/win32/browseui/CAutoComplete.cpp:2107]: @ThFabba said `MultiThreadTest` (https://jira.reactos.org/secure/attachment/59567/MultiThreadTest.zip) is alright. This PR is based on this `MultiThreadTest`. @ThFabba might have a sight of this problem.
